nilearn/glm/tests/test_utils.py

```diff
@@ -31,12 +37,82 @@ def test_full_rank():
+    cdf_val = sps.f.cdf(fval, 42, 1e10)
+    p_val = np.array(np.minimum(np.maximum(p_val, 1.e-300), 1. - 1.e-16))
+    cdf_val = np.array(np.minimum(np.maximum(cdf_val, 1.e-300), 1. - 1.e-16))
+    z_val_sf = norm.isf(p_val)
+    z_val_cdf = norm.ppf(cdf_val)
+    z_val = np.zeros(p_val.size)
+    z_val[np.atleast_1d(z_val_sf < 0)] = z_val_cdf[z_val_sf < 0]
+    z_val[np.atleast_1d(z_val_sf >= 0)] = z_val_sf[z_val_sf >= 0]
+    assert_array_almost_equal(z_score(p_val, cdf_val), z_val)
     # check the numerical precision
-    for p in [1.e-250, 1 - 1.e-16]:
-        assert_array_almost_equal(z_score(p), norm.isf(p))
-    assert_array_almost_equal(z_score(np.float32(1.e-100)), norm.isf(1.e-300))
+    for t in [33.75, -8.3]:
+        p = sps.t.sf(t, 1e10)
+        cdf = sps.t.cdf(t, 1e10)
+        z_sf = norm.isf(p)
+        z_cdf = norm.ppf(cdf)
+        if p <= .5:
+            z = z_sf
+        else:
+            z = z_cdf
+        assert_array_almost_equal(z_score(p, cdf), z)
+    assert_array_almost_equal(z_score(np.float32(1.e-100),
+                                      1-np.float32(1.e-100)),
+                              norm.isf(1.e-300))
+
+
+def test_z_score_opposite_contrast():
+    fmri, mask = generate_fake_fmri(shape=(50, 20, 50), length=96)
+
+    nifti_masker = NiftiMasker(mask_img=mask)
+    data = nifti_masker.fit_transform(fmri)
+
+    frametimes = np.linspace(0, (96 - 1) * 2, 96)
+
+    for i in [0, 20]:
+        design_matrix = make_first_level_design_matrix(
+            frametimes, hrf_model='spm',
+            add_regs=np.array(data[:, i]).reshape(-1, 1))
+        c1 = np.array([1] + [0] * (design_matrix.shape[1] - 1))
+        c2 = np.array([0] + [1] + [0] * (design_matrix.shape[1] - 2))
+        contrasts = {'seed1 - seed2': c1 - c2, 'seed2 - seed1': c2 - c1}
+        fmri_glm = FirstLevelModel(t_r=2., 
+                                   noise_model='ar1', 
+                                   standardize=False, 
+                                   hrf_model='spm', 
+                                   drift_model='cosine')
+        fmri_glm.fit(fmri, design_matrices=design_matrix)
+        z_map_seed1_vs_seed2 = fmri_glm.compute_contrast(
+            contrasts['seed1 - seed2'], output_type='z_score')
+        z_map_seed2_vs_seed1 = fmri_glm.compute_contrast(
+            contrasts['seed2 - seed1'], output_type='z_score')
+
+        print(z_map_seed1_vs_seed2.get_data().min())
+        print(-z_map_seed2_vs_seed1.get_data().max())
+        npt.assert_almost_equal(z_map_seed1_vs_seed2.get_data().min(),
+                                -z_map_seed2_vs_seed1.get_data().max(),
+                                decimal=10)
+        print(z_map_seed1_vs_seed2.get_data().max())
+        print(-z_map_seed2_vs_seed1.get_data().min())
+        npt.assert_almost_equal(z_map_seed1_vs_seed2.get_data().max(),
+                                -z_map_seed2_vs_seed1.get_data().min(),
+                                decimal=10)
```

Comment:
Note sure what is tested here: since the data are pure noise you won't have extreme z_values on one side or the other, will you ?

Reply:
This function is testing the numerical accuracy for a range of z values and it was raising an error for more extreme ones, but not anymore. This problem starts for |z| > 7 and it was visible in the spm_auditory data, as described in the original issue.

Reply:
But here, you may not encounter values of |z| greater than 7 ?

Reply:
Yes, we do. This is the log obtained from the prints in lines 125, 126, 130 and 131:

nilearn/glm/tests/test_utils.py::test_z_score_opposite_contrast 

-3.684085191397902
-3.684085191397902

3.925253566151492
3.925253566151492

-3.662796551676404
-3.662796551676404

37.0470962993612
37.0470962993612

Reply:
Can someone revise this PR? Thanks! :) @bthirion @KamalakerDadi @jeromedockes @NicolasGensollen

Reply:
> Yes, we do

can you add an assertion to make sure it will always be the case? (eg if the fake data generator is modified)

Reply:
so this test fails with the current nilearn but passes after merging this PR, is that correct?

Reply:
Yes. With current nilearn the behavior should be as demonstrated [here](https://gist.github.com/mwegrzyn/9349842ee2abbff05df735a595bb239a) qualitatively.

Reply:
>can you add an assertion to make sure it will always be the case? 

It is already the case, no? ```assert_almost_equal```

Reply:
I have the same question.

Reply:
sorry I meant to assert @bthirion 's comment: assert there is indeed z-values greater than 7